### ISSUE TITLE
twisted.web.static.File should allow customizing the ForbiddenResource

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6951]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#6951|
|Type|enhancement|
|Created|2014-02-08 11:02:02Z|
|Branch|https://github.com/twisted/twisted/tree/static-file-custom-forbidden-6951-2|

Right now, we have this code which makes it harder for implementation to create their own  custom forbidden resource.

```
        try:
            fileForReading = self.openForReading()
        except IOError, e:
            import errno
            if e[0] == errno.EACCES:
                return resource.ForbiddenResource().render(request)
            else:
                raise
```

Maybe this can be implemented in the same as childNotFound resource

```
childNotFound = resource.NoResource("File not found.")
```

Attachments:

* [6951-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d3c/d3cce3cd96bd14201115ac32a158e5c495853c5a/d454fb2379fb49192d2fdc0bf20872f6b686250d.diff) (1267 bytes) - added by adiroiban on 2014-02-08 11:23:57Z - 
* [6951-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d3c/d3cce3cd96bd14201115ac32a158e5c495853c5a/e2b7b1d2d7d5311a28b66ba2e024847eedebf518.diff) (2740 bytes) - added by adiroiban on 2014-02-08 13:13:27Z - 
* [6951-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d3c/d3cce3cd96bd14201115ac32a158e5c495853c5a/fe6123620437fea4538cd813df6cc61f80047b5b.diff) (3158 bytes) - added by adiroiban on 2014-03-03 13:11:05Z - 
* [6951-4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d3c/d3cce3cd96bd14201115ac32a158e5c495853c5a/5f065c3e304c1e8aee22c621568704d00c1c6560.diff) (2827 bytes) - added by adiroiban on 2014-11-10 18:04:06Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6951 6951
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_static_file_custom_forbidden_6951_2 branches/static-file-custom-forbidden-6951-2
branch_author__glyph__rwall glyph, rwall
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1391857322000000 1391857322000000
changetime__1424126948005306 1424126948005306
version__None None
owner__adiroiban adiroiban
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban removed owner|
|-|-|

I have attached an initial patch.

It is as simple code and I have not created a dedicated test as I think that the current test_forbiddenResource should cover this part on Unix.

Please let me know if a dedicated test is required.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @adiroiban|
|-|-|

Thanks.  This looks like a nice feature.  Can you add some unit tests?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @exarkun|
|-|-|

Thanks for the review.

I have attached a new diff which includes a unit test.

It is a bit of duplicate from test_forbiddenResource , but it mocks the openForReading method so that we can have this test run on Windows.

It is more of an integration test since it checks full usage of the new custom 'forbidden' resource.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

Just thought I'd mention that I did some related work in ticket:3765#comment:7 where I started writing a brand new set of static file rendering components. In that branch I added a pathNotFound initialiser argument which allows you to customise the behaviour without subclassing ([branches/static-path-3765/twisted/web/static.py#L1125).](branches/static-path-3765/twisted/web/static.py#L1125).)

Not sure when I'm going to get round to finishing it though.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj commented|
|-|-|

(In [41775]) Branching to 'static-file-custom-forbidden-6951'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj set owner to @adiroiban<br>@wallrj set status to `new`|
|-|-|

Thanks Adi,

This does look useful. Surprised noone needed this before.

Notes:

As in my previous review, I note that the git a / b prefixes make the patch a
little difficult to work with ...especially with the automated tools used by
some twisted developers.

[richard@zorin static-file-custom-forbidden-6951]$ patch -p0 < ../../6951-2.diff
can't find file to patch at input line 5
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/twisted/web/static.py b/twisted/web/static.py
|index a8932e6..3ff2d7a 100644
|--- a/twisted/web/static.py
|+++ b/twisted/web/static.py
--------------------------
File to patch: ^C
[richard@zorin static-file-custom-forbidden-6951]$ patch -p1 < ../../6951-2.diff
patching file twisted/web/static.py
patching file twisted/web/test/test_static.py
patching file twisted/web/topfiles/6951.misc

Patch applied cleanly though.

Points:
 1. twisted/web/topfiles/6951.misc
    1. This seems like a feature to me. Please add a .feature newsfile with a description
 2. twisted/web/test/test_static.py
    1. CustomForbiddenResource
       1. The fact that you have to create a static.File subclass seems ugly.
       2. Why not add a new "forbiddenResourceFactory" parameter to
          static.File.__init__? That way people can use this feature without
          having to create pointless subclasses.
       3. It also means that static.File._forbidden can be private -- rather than expanding the public API
       4. It should also make the test a little simpler.
    2. Add a test to demonstrate the default and overridden value of static.File.forbidden (and another showing that it can be overridden if you agree with the comment above)
    3. See source:branches/static-path-3765/twisted/web/test/test_static.py@39863#L1701 for an example of what I mean.
 3. twisted/web/static.py
    1. See suggested changes to static.File constructor arguments above.


Please address or answer the numbered points above and resubmit a new patch against trunk for review.

Find me on #twisted-dev if you'd like to discuss.

(once again, ignore the branch I created, that was premature)

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban removed owner|
|-|-|

   As in my previous review, I note that the git a / b prefixes make the
   patch a
   little difficult to work with ...especially with the automated tools used
   by
   some twisted developers.

Sorry for that. The new diff should be without prefix.

   Points:
    1. twisted/web/topfiles/6951.misc
       1. This seems like a feature to me. Please add a .feature newsfile
   with a description

Done.

    2. twisted/web/test/test_static.py
       1. CustomForbiddenResource
          1. The fact that you have to create a static.File subclass seems
   ugly.
          2. Why not add a new "forbiddenResourceFactory" parameter to
             static.File.__init__? That way people can use this feature
   without
             having to create pointless subclasses.
          3. It also means that static.File._forbidden can be private --
   rather than expanding the public API
          4. It should also make the test a little simpler.

I used the same design as childNotFound.

As in [#6933](https://github.com/twisted/twisted/issues/6933) , [#6972](https://github.com/twisted/twisted/issues/6972) and maybe other ticket, twisted web has many things
which requires customization.
Instead of adding new arguments to File.__init__ maybe we can define
some configuration object and pass it around.

Let me know if you want to have this in __init__ ... but then it would need
a new ticket to also fix childNotFound.

I will open a discussion on twisted mailing list to gather feedback for
configuration options.



       2. Add a test to demonstrate the default and overridden value of
   static.File.forbidden (and another showing that it can be overridden if
   you agree with the comment above)
       3. See source:branches/static-
   path-3765/twisted/web/test/test_static.py@39863#L1701 for an example of
   what I mean.

Not sure if this works as on is a class and another is an instance.
I have added an ugly assertIsInstance assertion.


    3. twisted/web/static.py
       1. See suggested changes to static.File constructor arguments above.

   Please address or answer the numbered points above and resubmit a new
   patch against trunk for review.


Please see attached patch and inline comments.
I  can go add add arguments to __init__ methods, but I think that this is
a generic problem and a separate ticket/discussion is required.

For this ticket it should use the same design as childNotFound.

Thanks for the review

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [43375]) Branching to static-file-custom-forbidden-6951-2.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [43376]) apply 6951-3.diff refs [#6951](https://github.com/twisted/twisted/issues/6951)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks again, adiroiban :).

  1. It looks like there are some new twistedchecker errors (one of which is clearly a typo...).
  2. on another ticket recently, exarkun opined that `test_something_somethingElse` naming is not really in the spirit of the coding standard, since no actual dynamic dispatching is going on.  For what it's worth, I was the one writing tests with that naming convention, but I think he's right, policy-wise :).
  3. I'm not too sure about the use of a `cvar` here.  Generally speaking `cvar`s are bad news; they're global, and can be overridden globally.  Also, the extension point is _treated_ like an `ivar`, (accessed via `self`, not a class) despite having a class-level and class-respected default.  how would you feel about having this be a straight `ivar ` instead, just to decrease (or rather, avoid further increasing) the amount of global state honored by this class?  If you feel that class-level is really the right solution for consistency, or for some use-case I haven't considered, feel free to say so, but new global state should require at least a little extra justification every time.

The buildbots are otherwise very green.  Please resubmit when you have some thoughts on how to address the above, this should be a good week to do that :-).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review.

* I have fixed the twistedchecker errors.

* I have  renamed the test to `test_forbiddenResourceCustomize` ... I still prefer `test_forbiddenResource_customize` as it is much easier to read that the test is targeted to the forbiddenResource attribute. 

* I don't want that forbidden resource to be a class member... I just find it easy to give it a default value as class member and easy to overwrite it in specialized classes. I have changed the code so that an instance member is used. Or maybe I can keep the previous code and just change the docstring. What do you think?


--------

I have attached a new diff based on latest trunk.

Thanks!

PS: Sorry for the delay, I was busy last weeks.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_14"></a>|@wsanchez commented|
|-|-|

A downside to the `cvar` -> `ivar` change is that now it's instantiating a new `ForbiddenResource` object every time a `File` is instantiated.  Further, it's doing so even when that `ForbiddenResource` is never used, so it's wasted churn.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez commented|
|-|-|

glyph: what do you think of a /private/ `cvar` and a property that vends that but sets (and then vends) an overriding `ivar`?

I think that addresses the global state by making it clear that the default/global resource is meant to be immutable.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_16"></a>|@wsanchez commented|
|-|-|

Quoting glyph from IRC:

  wsanchez: sorry, don't get too hung up on the cvar thing
  wsanchez: we just have a history of polluting class scope with way too many default values because they just happen to be immutable even though they shouldn't be visible on the class
  wsanchez: it's more that I really want to make sure that issue is *raised* every time we add a @cvar default for something, not necessarily that we never do that, just that it shouldn't be the go-to default any more

adiroiban: I suggest going back to patch 3's `cvar` usage.

It seems like some documentation should discourage stepping on class variables due to the global state change problem, and setting overriding instance variables instead.  But that doesn't seem like something to fix here, as it's a best practices thing not specific to this one case.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review.

6951-3.diff was already applied as HEAD of static-file-custom-forbidden-6951-2 

I have fixed the variable names in tests as done in  6951-4.diff.

Please take another look.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_19"></a>|@wsanchez set owner to @adiroiban|
|-|-|

Cool.  Test look good; I like the rendering of a marker.  I'm noticing that there are no corresponding tests for `childNotFound`.  That's not exactly this ticket, but perhaps you'd be game to add those opportunistically, since it's probably mostly identical.

OK to merge, but you'll need to camel-case some symbols to clear up a pair of new `twistedchecker` items… looks like you already fixed those… rebuilding… yup, good.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban removed owner|
|-|-|

I have added a test for childNotFound and also updated the docstrings to talk about class members.

Please take a look at latest changes.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_21"></a>|@wsanchez set owner to @adiroiban|
|-|-|

great thanks

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_22"></a>|@adiroiban set status to `closed`|
|-|-|

(In [43934]) Merging static-file-custom-forbidden-6951-2: Allow custom resource for forbidden response.

Author: adiroiban
Reviewer: rwall, wsanchez, glyph
Fixes: [#6951](https://github.com/twisted/twisted/issues/6951)

twisted.web.static.File allows defining a custom resource for rendering forbidden pages.
